File: receipt-analyzer-service/src/main/java/pl/madzierski/daniel/app/receipt/revision/model/GetRevisionResponse.java
package pl.madzierski.daniel.app.receipt.revision.model;

import pl.madzierski.daniel.app.file_group.file.FileEntity;
import pl.madzierski.daniel.app.receipt.revision.ReceiptRevisionEntity;
import pl.madzierski.daniel.app.receipt.revision.item.ItemEntity;
import pl.madzierski.daniel.app.receipt.scan_resolver.ReceiptResolverStrategyType;

import java.time.LocalDateTime;
import java.util.Set;
import java.util.stream.Collectors;

public record GetRevisionResponse(
        String id,
        ReceiptResolverStrategyType resolver,
        LocalDateTime createdDate,
        String brand,
        Double totalPrice,
        String payingDate,
        String address,
        Boolean isPreferredRevision,
        Boolean isCorrect,
        Set<ItemResponse> items
) {

    public static GetRevisionResponse revisionMapper(ReceiptRevisionEntity revisionEntity) {
        if (revisionEntity == null) {
            return null;
        }

        Set<ItemResponse> mappedItems = revisionEntity.getItems().stream()
                .map(GetRevisionResponse::itemMapper)
                .collect(Collectors.toSet());

        return new GetRevisionResponse(
                revisionEntity.getId(),
                revisionEntity.getResolver(),
                revisionEntity.getCreatedDate(),
                revisionEntity.getBrand(),
                revisionEntity.getTotalPrice(),
                revisionEntity.getPayingDate(),
                revisionEntity.getAddress(),
                revisionEntity.getIsPreferredRevision(),
                revisionEntity.getIsCorrect(),
                mappedItems
        );
    }

    public static FileResponse fileMapper(FileEntity file) {
        return new FileResponse(
                file.getId(),
                file.getPath(),
                file.getRawData()
        );
    }

    public static ItemResponse itemMapper(ItemEntity itemEntity) {
        return new ItemResponse(
                itemEntity.getId(),
                itemEntity.getName(),
                itemEntity.getAmount(),
                itemEntity.getUnitPrice(),
                itemEntity.getDiscount(),
                itemEntity.getTotalPrice(),
                itemEntity.getPosition()
        );
    }

    public record ItemResponse(
            String id,
            String name,
            Double amount,
            Double unitPrice,
            Double discount,
            Double totalPrice,
            Integer position
    ) {
    }

    public record FileResponse(
            String id,
            String path,
            String rawData
    ) {
    }
}

File: receipt-analyzer-service/src/main/kotlin/pl/madzierski/daniel/app/receipt/revision/model/GetRevisionResponse.kt
package pl.madzierski.daniel.app.receipt.revision.model

import pl.madzierski.daniel.app.receipt.revision.ReceiptRevisionEntity
import pl.madzierski.daniel.app.receipt.revision.ScanResolver
import pl.madzierski.daniel.app.receipt.revision.item.ItemEntity
import pl.madzierski.daniel.app.file_group.file.FileEntity
import java.time.LocalDateTime


data class GetRevisionResponse(
    var id: String?,
    var resolver: ScanResolver?,
    val createdDate: LocalDateTime?,
    val brand: String?,
    val totalPrice: Double?,
    val payingDate: String?,
    val address: String?,
    val isPreferredRevision: Boolean?,
    val isCorrect: Boolean?,
    val items: Set<ItemResponse>?
) {

    data class ItemResponse(
        val id: String?,
        val name: String?,
        val ptu: String?,
        val amount: Double?,
        val unitPrice: Double?,
        val discount: Double?,
        val totalPrice: Double?,
        val position: Int?,
    )

    data class FileResponse(
        val id: String?, val path: String?, val rawData: String?
    )

    companion object {

        fun revisionMapper(revisionEntity: ReceiptRevisionEntity?): GetRevisionResponse? {
            if (revisionEntity != null) {
                return GetRevisionResponse(
                    revisionEntity.id,
                    revisionEntity.resolver,
                    revisionEntity.createdDate,
                    revisionEntity.brand,
                    revisionEntity.totalPrice,
                    revisionEntity.payingDate,
                    revisionEntity.address,
                    revisionEntity.isPreferredRevision,
                    revisionEntity.isCorrect,
                    revisionEntity.items.mapTo(mutableSetOf()) { itemMapper(it) }
                )
            }
            return null
        }

        fun fileMapper(file: FileEntity) = FileResponse(
            file.id, file.path, file.rawData
        )

        fun itemMapper(itemEntity: ItemEntity): ItemResponse = ItemResponse(
            itemEntity.id,
            itemEntity.name,
            itemEntity.ptu,
            itemEntity.amount,
            itemEntity.unitPrice,
            itemEntity.discount,
            itemEntity.totalPrice,
            itemEntity.position
        )

    }
}
